File: local/src/lib/gcounter.js
export class GCounter {
    #map = new Map();
    #id;

    constructor(id) {
        this.#id = id;
        this.#map[id] = 0;
    }

    // Set key
    setKey(id) {
        this.#id = id;
        if (this.#map[this.#id] == undefined) this.#map[this.#id] = 0;
    }

    // Get map
    get() {
        const res = this.#map;
        return res;
    }

    // Read local counter value
    local() {
        return this.#map.get(this.#id);
    }

    // Read counter value
    read() {
        let res = 0;
        for (const [id, value] of Object.entries(this.#map)) {
            res += value;
        }
        return res;
    }

    // Increment value
    inc(tosum = 1) {
        this.#map[this.#id] += tosum;
    }

    // Merge counters
    join(gcounter) {
        for (const [id, value] of Object.entries(gcounter.get())) {
            if (!this.#map.has(id)) this.#map[id] = 0;
            this.#map[id] = Math.max(this.#map[id], value);
        }
    }

    fromJSON(counter) {
        this.#map = new Map();
        for (const [key, value] of counter) {
            this.#map[key] = value;
        }
    }

    toJSON() {
        const res = []; 
        for (const [key, value] of Object.entries(this.#map)) {
            console.log(key, value);
            res.push([key, value]);
        }
        
        return res;
    }
}

File: cloud/src/main/java/server/model/GCounter.java
package server.model;

import java.util.HashMap;
import java.util.Map;

public class GCounter {
    private Map<String, Integer> map = new HashMap<>();
    private String id;

    // Constructor with id
    public GCounter(String id) {
        this.id = id;
        this.map.put(id, 0);
    }

    // Constructor for creating a copy of the counter
    public GCounter(GCounter counter) {
        this.map = new HashMap<>(counter.map);
        this.id = counter.id;
    }

    // Set key
    public void setKey(String id) {
        this.id = id;
        this.map.put(id, 0);
    }

    // Get map
    public Map<String, Integer> get() {
        return map;
    }

    // Read local counter value
    public int local() {
        return map.get(id);
    }

    // Read counter value
    public int read() {
        int res = 0;
        for (int value : map.values()) {
            res += value;
        }
        return res;
    }

    // Increment value
    public void inc(int toSum) {
        map.put(id, map.get(id) + toSum);
    }

    // Merge counters
    public void join(GCounter gCounter) {
        for (Map.Entry<String, Integer> entry : gCounter.map.entrySet()) {
            String entryId = entry.getKey();
            int entryValue = entry.getValue();
            map.put(entryId, Math.max(map.getOrDefault(entryId, 0), entryValue));
        }
    }
}